rust-symcrypt/src/gcm.rs

```diff
@@ -0,0 +1,377 @@
+        let mut instance = Box::new(GcmState {
+            state: symcrypt_sys::SYMCRYPT_GCM_STATE::default(),
+            expanded_key: symcrypt_sys::SYMCRYPT_GCM_EXPANDED_KEY::default()
+        });
+        unsafe {
+            expand_key(key, &mut instance.expanded_key)?;
+            symcrypt_sys::SymCryptGcmInit(
+                &mut instance.state,
+                &instance.expanded_key,
+                nonce.as_ptr(),
+                nonce.len() as symcrypt_sys::SIZE_T,
+            );
+            Ok(instance)
+        }
+    }
+
+    pub fn auth(&mut self, auth_data: &[u8]) {
+        unsafe {
+            symcrypt_sys::SymCryptGcmAuthPart(
+                &mut self.state,
+                auth_data.as_ptr(),
+                auth_data.len() as symcrypt_sys::SIZE_T,
+            );
+        }
+    }
+
+    pub fn encrypt_part(&mut self, data: &[u8]) -> Vec<u8> {
+        unsafe {
+            let mut dst: Vec<u8> = vec![0u8; data.len()];
+            symcrypt_sys::SymCryptGcmEncryptPart(
+                &mut self.state,
+                data.as_ptr(),
+                dst.as_mut_ptr(),
+                dst.len() as symcrypt_sys::SIZE_T,
+            );
+            dst
+        }
+    }
+
+    pub fn encrypt_final(&mut self, tag: &mut [u8]) {
+        unsafe {
+            symcrypt_sys::SymCryptGcmEncryptFinal(
+                &mut self.state,
+                tag.as_mut_ptr(),
+                tag.len() as symcrypt_sys::SIZE_T,
+            );
+        }
+    }
+
+    pub fn decrypt_part(&mut self, data: &[u8]) -> Vec<u8> {
+        let mut dst: Vec<u8> = vec![0u8; data.len()];
+        unsafe {
+            symcrypt_sys::SymCryptGcmDecryptPart(
+                &mut self.state,
+                data.as_ptr(),
+                dst.as_mut_ptr(),
+                data.len() as symcrypt_sys::SIZE_T,
+            );
+            dst
+        }
+    }
+
+    pub fn decrypt_final(&mut self, tag: &[u8]) -> Result<(), SymCryptError> {
+        unsafe {
+            match symcrypt_sys::SymCryptGcmDecryptFinal(
+                &mut self.state,
+                tag.as_ptr(),
+                tag.len() as symcrypt_sys::SIZE_T,
+            ) {
+                symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => Ok(()),
+                err => Err(err.into()), 
+            }
+        }
+    }
+}
+
+impl Drop for GcmState {
+    fn drop(&mut self) {
+        unsafe {
+            symcrypt_sys::SymCryptWipe(
+                ptr::addr_of_mut!(self.state) as *mut c_void,
+                mem::size_of_val(&mut self.state) as symcrypt_sys::SIZE_T,
+            );
+            symcrypt_sys::SymCryptWipe(
+                ptr::addr_of_mut!(self.expanded_key) as *mut c_void,
+                mem::size_of_val(&mut self.expanded_key) as symcrypt_sys::SIZE_T,
+            );
+        }
+    }
+}
+
+pub fn validate_gcm_parameters(
+    block_cipher: symcrypt_sys::PCSYMCRYPT_BLOCKCIPHER,
```

Comment:
>  block_cipher: symcrypt_sys::PCSYMCRYPT_BLOCKCIPHER, [](http://example.com/codeflow?start=2&length=53)

This should be a rust enum. No callers should need to know about any symcrypt_sys information.